### DISCUSSION TITLE
The right way to select and update with AsyncSession

### DISCUSSION THREAD

**Discussion by mockingjet** [sqlalchemy/sqlalchemy#6630] [Usage Questions] [answered]
The following code is a demo of how I select a row and update it.
But I do think it is **not** straightforward for me. (Although it works.)

Is this the right to way to update a row? 
Since select query and update statement are executed seperately, how to make sure both work on the same row?

And I found I don't need to **db.resfresh** that object, it will return the updated one, and I don't know why ...
( It may be purposely designed this way?)
 
Hope someone can help explain this. Thanks!

```py
async def modify(
        self,
        db: AsyncSession,
        user_id: str,
        payload: Union[UserProfile, UserPersonProfile],
    ) -> UserProfileModel:
        """modify user profile partially"""

       # Normally when we try to update a row, we will find out if this row exists
        query = sa.select(UserProfileModel).where(UserProfileModel.id == user_id)
        result = await db.execute(query)
        profile = result.scalars().first()
        
        # if not, raise exception
        if not profile:
            raise HTTPException(400, detail=["profile not found"])
        
        # here is my question
        stmt = (
            sa.update(UserProfileModel)
            .where(UserProfileModel.id == user_id)
            .values(**payload.dict())
        )
        result = await db.execute(stmt)
        
        # why I don't need this?
        # db.refresh(profile)

        return profile
```

**Comment by mockingjet** [comment]
Another question is will db.execute() auto commit?
Profile data is saved on database without calling await db.commit().

**Reply by CaselIT** [reply]
That depends on the configuration of the dbapi. sqlalchemy does not commit by default

**Comment by CaselIT** [comment, marked as answer]
Hi,

That's the expected behaviour, since when using an orm enabled update the object in the session are kept in sync with the db.

That said, it's a bit unusual to use an `update` when the orm object is available, since one would usually modify the properties on the object directly and then commit the changes to the db. See this part of the tutorial https://docs.sqlalchemy.org/en/14/tutorial/orm_data_manipulation.html

As a side-nete, assuming `UserProfileModel.id` is the primary key of the model, you may also use the `session.get` method:
```py
profile = await session.get(UserProfileModel, user_id)
```

**Reply by mockingjet** [reply]
>  ... since one would usually modify the properties on the object directly and then commit the changes to the db.

This part is true. But it doesn't support **partially update** !!!

If update it like ```profile.email = "xxx@gmail.com"```, I have to check ```if payload.email is not None``` 

What I can make sure is ```payload``` has attributes only included in ```UserProfileModel```

**Reply by mockingjet** [reply]
Yeah, I tried ```await session.get``` yesterday, and it worked the same.
